nat-lab/tests/test_batching.py

```diff
@@ -0,0 +1,207 @@
+        alpha, beta, gamma, delta = env.nodes
+
+        clients = [alpha_client, beta_client, gamma_client, delta_client]
+        peer_nodes = [alpha, beta, gamma, delta]
+
+        await asyncio.gather(*(
+            client.wait_for_state_peer(
+                peer_node.public_key, [State.Connected], [PathType.Direct]
+            )
+            for client, peer_node in product(clients, peer_nodes)
+            if not client.is_node(peer_node)
+        ))
+
+        # Capturing the traffic can be done on nodes or gateways
+        # On gateways there's a guarantee that the traffic has arrived
+        # however we can't decrypt it easily.
+        # On nodes there's no guarantee the packet left the network adapter
+        # but we can inspect the packet before it hit the TUN device.
+        # Here we capture on both
+        gateways = [DOCKER_GW_MAP[param.connection_tag] for param in setup_params]
+        gateway_container_names = [container_id(conn_tag) for conn_tag in gateways]
+        conns = [client.get_connection() for client in env.clients]
+        node_container_names = [
+            conn.container_name()
+            for conn in conns
+            if isinstance(conn, DockerConnection)
+        ]
+
+        container_names = gateway_container_names + node_container_names
+        # TODO: assert these are docker connections
+        print("Will capture batching on containers: ", container_names)
+        cnodes = zip(env.clients, env.nodes, [c.connection for c in env.connections])
+
+        # the plan is to misalign the peers for better batcing observation. WIthout misalignment
+        # peers will form a meshnet almost immediately, already achieving the desired batching effect.
+        for client in env.clients:
+            await client.stop_device()
+
+        tasks = []
+        for name in container_names:
+            task = asyncio.create_task(
+                capture_traffic(
+                    name,
+                    capture_duration,
+                )
+            )
+            tasks.append(task)
+
+        api = env.api
+
+        # misalign the peers by sleeping some before starting each node again
+        for client, node, _ in cnodes:
+            await asyncio.sleep(6)
+            await client.simple_start()
+            await client.set_meshmap(api.get_meshmap(node.id))
+
+        await asyncio.gather(*(
+            client.wait_for_state_peer(
+                peer_node.public_key, [State.Connected], [PathType.Direct]
+            )
+            for client, peer_node in product(clients, peer_nodes)
+            if not client.is_node(peer_node)
+        ))
+
+        pcap_paths = await asyncio.gather(*tasks)
+
+        # pylint: disable=consider-using-set-comprehension
+        pyro5_ports = list(set([client.get_proxy_port() for client in env.clients]))
+        assert len(pyro5_ports) == 1
+        pyro5_port = pyro5_ports[0]
```

Comment:
Maybe you can just take this [0] immediately, I think the error will be quite obvious if there is none (or is it a check for multiple ports?)

Reply:
I am not yet sure how Pyro5 works and based on my understanding it chooses mapped Docker's ports. My understanding is that it might be different for each client, however from the observation I see that it's always the same.

Making those filters unique based on each `pyro5_ports` requires extra work so I just went the lazy way - assert they're all equal and grab the first one

Reply:
The ports are not all the same when running in CI, but will be the same on the container side (but not host side) when running locally